compiler/rustc_hir_typeck/src/cast.rs

```diff
@@ -828,13 +828,18 @@ impl<'a, 'tcx> CastCheck<'tcx> {
             (Some(PointerKind::VTable(src_tty)), Some(PointerKind::VTable(dst_tty))) => {
                 match (src_tty.principal(), dst_tty.principal()) {
                     // A<dyn Trait + Auto> -> B<dyn Trait' + Auto'>. need to make sure
-                    // - traits are the same & have the same generic arguments
+                    // - traits are the same
+                    // - traits have the same generic arguments
                     // - Auto' is a subset of Auto
-                    //
-                    // This is checked by checking `dyn Trait + Auto + 'erased: Unsize<dyn Trait' + Auto' + 'erased>`.
-                    (Some(_), Some(_)) => {
+                    (Some(src_principal), Some(dst_principal)) => {
                         let tcx = fcx.tcx;
 
+                        // Check that the traits are actually the same
+                        // (this is required as the `Unsize` check below would allow upcasting, etc)
```

Comment:
we'll want to allow that in the future, right?

Reply:
I'm pretty sure upcasting cast is handled somewhere else and this is only for "type change without changing the value" kind of thing. Will have to double check this.,

Reply:
Just double checked, upcasting goes through `coerce_unsized` and is thus unaffected. Will add a comment in that regard.